From ext-phil.2.carmody@nokia.com  Tue Dec 22 11:52:59 2009
From: Phil Carmody <ext-phil.2.carmody@nokia.com>
Date: Mon, 14 Dec 2009 20:28:12 +0200
Subject: driver core: Prevent reference to freed memory on error path
To: gregkh@suse.de, kay.sievers@vrfy.org
Cc: linux-kernel@vger.kernel.org
Message-ID: <1260815292-18538-1-git-send-email-ext-phil.2.carmody@nokia.com>


priv is drv->p. So only free drv->p after we've finished using priv.

Found using a static code analysis tool

Signed-off-by: Phil Carmody <ext-phil.2.carmody@nokia.com>
Signed-off-by: Greg Kroah-Hartman <gregkh@suse.de>

---
 drivers/base/bus.c |    2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

--- a/drivers/base/bus.c
+++ b/drivers/base/bus.c
@@ -703,9 +703,9 @@ int bus_add_driver(struct device_driver 
 	return 0;
 
 out_unregister:
+	kobject_put(&priv->kobj);
 	kfree(drv->p);
 	drv->p = NULL;
-	kobject_put(&priv->kobj);
 out_put_bus:
 	bus_put(bus);
 	return error;
